solr/core/src/java/org/apache/solr/search/grouping/distributed/responseprocessor/SkipSecondStepSearchGroupShardResponseProcessor.java

```diff
@@ -0,0 +1,116 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.search.grouping.distributed.responseprocessor;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.lucene.search.Sort;
+import org.apache.lucene.search.TotalHits;
+import org.apache.lucene.search.grouping.GroupDocs;
+import org.apache.lucene.search.grouping.SearchGroup;
+import org.apache.lucene.search.grouping.TopGroups;
+import org.apache.lucene.util.BytesRef;
+import org.apache.solr.handler.component.ResponseBuilder;
+import org.apache.solr.handler.component.ShardDoc;
+import org.apache.solr.handler.component.ShardRequest;
+import org.apache.solr.handler.component.ShardResponse;
+import org.apache.solr.search.SolrIndexSearcher;
+import org.apache.solr.search.grouping.GroupingSpecification;
+import org.apache.solr.search.grouping.distributed.shardresultserializer.SearchGroupsResultTransformer;
+
+public class SkipSecondStepSearchGroupShardResponseProcessor extends SearchGroupShardResponseProcessor {
+
+  @Override
+  protected SearchGroupsResultTransformer newSearchGroupsResultTransformer(SolrIndexSearcher solrIndexSearcher) {
+    return new SearchGroupsResultTransformer.SkipSecondStepSearchResultResultTransformer(solrIndexSearcher);
+  }
+
+  @Override
+  protected SearchGroupsContainer newSearchGroupsContainer(ResponseBuilder rb) {
+    return new SkipSecondStepSearchGroupsContainer(rb.getGroupingSpec().getFields());
+  }
+
+  @Override
+  public void process(ResponseBuilder rb, ShardRequest shardRequest) {
+    super.process(rb, shardRequest);
+    TopGroupsShardResponseProcessor.fillResultIds(rb);
+  }
+
+  protected static class SkipSecondStepSearchGroupsContainer extends SearchGroupsContainer {
+
+    private final Map<Object, String> docIdToShard = new HashMap<>();
+
+    public SkipSecondStepSearchGroupsContainer(String[] fields) {
+      super(fields);
+    }
+
+    @Override
+    public void addSearchGroups(ShardResponse srsp, String field, Collection<SearchGroup<BytesRef>> searchGroups) {
+      super.addSearchGroups(srsp, field, searchGroups);
+      for (SearchGroup<BytesRef> searchGroup : searchGroups) {
+        assert(srsp.getShard() != null);
+        docIdToShard.put(searchGroup.topDocSolrId, srsp.getShard());
+      }
+    }
+
+    @Override
+    public void addMergedSearchGroups(ResponseBuilder rb, String groupField, Collection<SearchGroup<BytesRef>> mergedTopGroups ) {
+      // TODO: add comment or javadoc re: why this method is overridden as a no-op
```

Comment:
The https://github.com/cpoerschke/lucene-solr/commit/d4af896d511b10b40263ee5383ed633609ff7ff1 commit turns this 'TODO' into method-level javadocs. Adding similar method-level javadocs for the `addSearchGroupToShards` method below made me wonder if
* (a) the `super` method call in it could be omitted, and
* (b) if emptiness of the relevant ResponseBuilder structures might be used as a way of determining that the 'second step' stage can be skipped?

Reply:
(a) good catch, `addSearchGroupToShard` only sets `searchGroupToShards` that seems to be used only here: 
 
https://github.com/apache/lucene-solr/blob/1d85cd783863f75cea133fb9c452302214165a4d/solr/core/src/java/org/apache/solr/search/grouping/distributed/requestfactory/TopGroupsShardRequestFactory.java#L69

and if we enable las vegas we are going to skip topgroups so we should be fine. 

(b) um I would still skip it explicit instead of relying on implicit signal that might change over the time..

Reply:
I removed `addSearchGroupToShard` and pushed here. Test are still successful, thanks!